---
output:
  html_document:
    df_print: paged
    toc: true
    number_sections: false
    toc_depth: 3
    toc_float:
      collapsed: false
      smooth_scroll: false
editor_options: 
  chunk_output_type: console
params:
  typ_map: typ_map
  typ_ped: typ_ped
  imp_map: imp_map
  imp_ped: imp_ped
  imp_info: imp_info
  info_cut: info_cut  
  sample: sample
  rwd: rwd
---
# `r params$sample`
Mitochdonrial SNP Imputation report of `r params$sample` generated on `r Sys.Date()`. Imputation was performed using the [MitoImpute](https://github.com/sjfandrews/MitoImpute) [IMPUTE2](http://mathgen.stats.ox.ac.uk/impute/impute_v2.html) pipeline.  
<br> 
```{r setup, echo=FALSE, warning =FALSE, message=FALSE}
knitr::opts_knit$set(root.dir = params$rwd)

library(tidyverse)
library(ggplot2)
library(ggforce)
library(HiMC); data(nodes)

## Tibble of SNPs used in Hi-MC and the assocaited haplogroup
HiMC <- tibble(
  himc = 'yes',
  pos = as.numeric(c('10115', '1018', '10398', '10400', '10550', '11177', '11251', '11719', '11947', '12007', '12308', '12414', '12705', '13263', '13368', '13506', '13708', '13789', '14178', '14318', '1438', '14470', '14560', '14668', '14766', '14905', '15043', '15326', '15452', '15535', '16111', '16189', '16271', '16362', '16390', '16391', '16391', '1719', '1736', '2092', '3505', '3552', '3594', '4580', '4769', '4883', '4917', '4977', '5178', '5442', '6371', '7028', '825', '8251', '8414', '8468', '8703', '9042', '9055', '9347', '9950')),
Haplogroup = c("L2", "L3", "K1", "M", "K", "B2", "JT", "R0", "W", "A2", "U", "N2", "R", "C", "T", "L2'3'4'6", "J", "L1", "L1", "C", "H2", "K", ".", "D4", "HV", "T", "N1a1b", "H2a2a", "JT", "B4b'd'e", "A2", "T1", "JT", "L4", "L2", "I", "I", "X2", "A", "D1", "W", "C", "L3'4", "V", "H2a", "M80'D", "T", "B2", "D ", "L0", "X ", "H", "L2'3'4'6", "N1a1b", "D4", "L2'3'4'6", "D2", "L0", "U8b", "L0", "B2"))

```

```{r readin, echo=FALSE, warning =FALSE, message=FALSE}
##  Read in - Typed only
#     Plink .ped files 
samp_typ.ped <- generate_snp_data(params$typ_map, params$typ_ped)

##  Read in - Imputed 
#     Plink .ped files 
samp_imp.ped <- generate_snp_data(params$imp_map, params$imp_ped) 
samp_imp.ped <- samp_imp.ped[,-c(grep("\\<189\\>", colnames(samp_imp.ped)), grep("\\<16183\\>", colnames(samp_imp.ped)))]
samp_imp.info <- read_delim(params$imp_info, delim = " ")

## Combine info scores
samp_imp.info <- samp_imp.info %>% 
  mutate(info_comb = ifelse(info_type0 == -1, info,info_type0 )) %>% 
  left_join(HiMC, by = c('position' = 'pos')) %>%
  mutate(himc = ifelse(is.na(himc), 'no', himc))

```

##Summary

SNPs in Reference Panel only (type 0): `r nrow(filter(samp_imp.info, type == 0))`


SNPs in Reference & Sample Panel only (type 2): `r nrow(filter(samp_imp.info, type == 2))`


SNPs in Sample Panel only (type 3): `r nrow(filter(samp_imp.info, type == 3))`


`r nrow(filter(samp_imp.info, info > params$info_cut))` out of `r nrow(samp_imp.info)` mtSNPs (`r nrow(filter(samp_imp.info, info > params$info_cut & himc == 'yes'))`/`r nrow(filter(samp_imp.info, himc == 'yes'))` Hi-MC mtSNPs) are retained using an Info threshold of `r params$info_cut`

<br>

**Table 1:** SNP-wise information 
```{r, echo=F, message = FALSE, warning=F}

samp_imp.info %>% 
  select(rs_id, position, a0, a1, exp_freq_a1, info, certainty, type, info_type0, concord_type0, r2_type0, himc, Haplogroup) %>% 
  as.data.frame()

```
<br>

##INFO Score
IMPUTE2 reports an information metric similar to r2, with values typically between 0 and 1 where values near 1 indicate that a SNP has been imputed with high certainty. For type 2 SNPs, IMPUTE2 masks the genotypes of one variant at a time in the study data (Panel 2), then imputes the masked genotypes with information from the reference data and nearby study variants.

```{r, echo=F, message = FALSE, warning=F, fig.width=7.5, fig.height=4.7, fig.align='center', fig.cap= 'Fig. 1: Info Score of imputed mtSNPs across the mitochondrial genome'}

ggplot(samp_imp.info, aes(x = position, y = info_comb, colour = as.factor(type))) + 
  geom_point() + theme_bw() + 
  labs(x = 'mtDNA position', y = 'Info Score') + geom_hline(yintercept = as.numeric(params$info_cut), colour = 'red', linetype = 2) +  
  guides(colour=guide_legend(title="SNP Type")) 

```
<br> 
```{r, echo=F, message = FALSE, warning=F, fig.width=7.5, fig.height=4.7, fig.align='center', fig.cap= 'Fig. 2: Info Score of imputed mtSNPs across the mitochondrial genome highlighting Hi-MC mtSNPs used in haplgroup assignment'}

ggplot(samp_imp.info, aes(x = position, y = info_comb, colour = as.factor(type), alpha = himc)) + geom_point() + theme_bw() + 
  labs(x = 'mtDNA position', y = 'Info Score') + geom_hline(yintercept = as.numeric(params$info_cut), colour = 'red', linetype = 2) + 
  guides(colour=guide_legend(title="SNP Type")) 

```
<br>

##Mitochondrial Haplogroups
[Hi-MC (Smieszek et al PeerJ 2018)](https://peerj.com/articles/5149/), a broad haplogroup classifier for European, African, and Native American mitochondrial haplogroups, is used to assign samples to macro haplogroups. Hi-MC was developed using a reduced mitochondrial phylogenetic tree targeting specific single nucleotide polymorphisms (SNPs) for genotyping.

```{r haplogorups, echo=FALSE, warning =FALSE, message=FALSE}
#   Typed Only
MTtyp.classifications <- HiMC::getClassifications(samp_typ.ped)

#   Imputed 
## remove sites with poor info scores (< 0.3)
rm.info <- filter(samp_imp.info, info > as.numeric(params$info_cut))
samp_imp.filtered <- samp_imp.ped[,colnames(samp_imp.ped) %in% c('Individual', rm.info$position)]

MTimp.classifications <- HiMC::getClassifications(samp_imp.filtered)

## Merge Samples
MT_haps <- MTtyp.classifications %>% 
  left_join(MTimp.classifications, by = 'Individual', suffix = c("_typ", "_imp")) %>% 
  as.tibble()
```

**Table 1:** Count of macro haplogroups in Sample pre- and post-imputation
```{r, echo=FALSE}

count(MT_haps, haplogroup_typ) %>% 
  full_join(count(MT_haps, haplogroup_imp), by = c('haplogroup_typ' = 'haplogroup_imp'), suffix = c('.typ', '.imp')) %>% 
  rename(Haplogroup = haplogroup_typ) %>% 
  arrange(Haplogroup)
  
```
<br>

### Alluvial Diagram
```{r haplogorups2, echo=FALSE, warning =FALSE, message=FALSE}

## Count pairs of haplogroups of imputed and WGS assignments
hap.match <- MT_haps %>%
  count(haplogroup_typ, haplogroup_imp) %>% 
  mutate(match = haplogroup_typ == haplogroup_imp) 

## Use ggforce to tidy data for geom_parallel Sets 
# Requires developmental version of ggforce
dat_ggforce <- hap.match  %>%
  gather_set_data(1:2) %>%       
  arrange(x,haplogroup_imp,desc(haplogroup_typ)) %>% 
  mutate(x = recode(x, 'haplogroup_imp' = '2', 'haplogroup_typ' = '0'))

```
<br>

```{r haplogorups3, echo=FALSE, warning =FALSE, message=FALSE, fig.width=7.5, fig.height=7.5, fig.align='center', fig.cap= 'Fig. 3: Haplogroup assignments pre- and post-imputation'}
ggplot(dat_ggforce, aes(x = x, id = id, split = y, value = n)) +
  geom_parallel_sets(aes(fill = match), alpha = 0.5, axis.width = 0.2) +
  geom_parallel_sets_labels(colour = 'black', angle = 0, size = 3) + 
  theme_classic() + theme(legend.position = 'bottom') + 
  theme(axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank(),
        text = element_text(size=12)) + 
  scale_x_discrete(labels=c("Typed", "Imputed")) + 
  labs(x = "Mitochondrial Haplogroups") + scale_fill_brewer(palette = 'Paired')

```








































